# Hmwk 1 Part 3 - Homer Kay

In [1]:
#Add packages
%matplotlib inline 
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
#From Scikit Learn
from sklearn import preprocessing
from sklearn.model_selection  import train_test_split, cross_val_score, KFold
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import roc_curve, auc, confusion_matrix, classification_report

In [2]:
# Checking Current Directory
%pwd

'/Users/homerkay/Desktop/6303 Predictive Models/HMWK 1/Notebooks'

In [3]:
# Change Directory
%cd /Users/homerkay/Desktop/6303 Predictive Models/HMWK 1

/Users/homerkay/Desktop/6303 Predictive Models/HMWK 1


## Reading in Dataset from URL

In [27]:

import pandas as pd
url = 'http://archive.ics.uci.edu/ml/machine-learning-databases/communities/communities.data'
crime = pd.read_csv(url, header=None, na_values=['?'])
crime.head(10)

#https://archive.ics.uci.edu/ml/datasets/Communities+and+Crime


0     1        2                    3    4     5     6     7     8     9    \
0    8   NaN      NaN         Lakewoodcity    1  0.19  0.33  0.02  0.90  0.12   
1   53   NaN      NaN          Tukwilacity    1  0.00  0.16  0.12  0.74  0.45   
2   24   NaN      NaN         Aberdeentown    1  0.00  0.42  0.49  0.56  0.17   
3   34   5.0  81440.0  Willingborotownship    1  0.04  0.77  1.00  0.08  0.12   
4   42  95.0   6096.0    Bethlehemtownship    1  0.01  0.55  0.02  0.95  0.09   
5    6   NaN      NaN    SouthPasadenacity    1  0.02  0.28  0.06  0.54  1.00   
6   44   7.0  41500.0          Lincolntown    1  0.01  0.39  0.00  0.98  0.06   
7    6   NaN      NaN            Selmacity    1  0.01  0.74  0.03  0.46  0.20   
8   21   NaN      NaN        Hendersoncity    1  0.03  0.34  0.20  0.84  0.02   
9   29   NaN      NaN          Claytoncity    1  0.01  0.40  0.06  0.87  0.30   

   ...    118   119   120   121   122  123  124   125   126   127  
0  ...   0.12  0.26  0.20  0.06  0.04  0.9  0.5  0.32  0.14  0.20  
1  ...   0.02  0.12  0.45   NaN   NaN  NaN  NaN  0.00   NaN  0.67  
2  ...   0.01  0.21  0.02   NaN   NaN  NaN  NaN  0.00   NaN  0.43  
3  ...   0.02  0.39  0.28   NaN   NaN  NaN  NaN  0.00   NaN  0.12  
4  ...   0.04  0.09  0.02   NaN   NaN  NaN  NaN  0.00   NaN  0.03  
5  ...   0.01  0.58  0.10   NaN   NaN  NaN  NaN  0.00   NaN  0.14  
6  ...   0.05  0.08  0.06   NaN   NaN  NaN  NaN  0.00   NaN  0.03  
7  ...   0.01  0.33  0.00   NaN   NaN  NaN  NaN  0.00   NaN  0.55  
8  ...   0.04  0.17  0.04   NaN   NaN  NaN  NaN  0.00   NaN  0.53  
9  ...   0.00  0.47  0.11   NaN   NaN  NaN  NaN  0.00   NaN  0.15  

[10 rows x 128 columns]

## Crime Target Variable

In [28]:
# Move Crime column (127).
# designate target variable name
targetName = 127
targetSeries = crime[targetName]
#remove target from current location and insert in collum 0
del crime[targetName]
crime.insert(0, targetName, targetSeries)
#reprint dataframe and see target is in position 0
crime.head(10)

127  0     1        2                    3    4     5     6     7     8    \
0  0.20    8   NaN      NaN         Lakewoodcity    1  0.19  0.33  0.02  0.90   
1  0.67   53   NaN      NaN          Tukwilacity    1  0.00  0.16  0.12  0.74   
2  0.43   24   NaN      NaN         Aberdeentown    1  0.00  0.42  0.49  0.56   
3  0.12   34   5.0  81440.0  Willingborotownship    1  0.04  0.77  1.00  0.08   
4  0.03   42  95.0   6096.0    Bethlehemtownship    1  0.01  0.55  0.02  0.95   
5  0.14    6   NaN      NaN    SouthPasadenacity    1  0.02  0.28  0.06  0.54   
6  0.03   44   7.0  41500.0          Lincolntown    1  0.01  0.39  0.00  0.98   
7  0.55    6   NaN      NaN            Selmacity    1  0.01  0.74  0.03  0.46   
8  0.53   21   NaN      NaN        Hendersoncity    1  0.03  0.34  0.20  0.84   
9  0.15   29   NaN      NaN          Claytoncity    1  0.01  0.40  0.06  0.87   

   ...    117   118   119   120   121   122  123  124   125   126  
0  ...   0.29  0.12  0.26  0.20  0.06  0.04  0.9  0.5  0.32  0.14  
1  ...    NaN  0.02  0.12  0.45   NaN   NaN  NaN  NaN  0.00   NaN  
2  ...    NaN  0.01  0.21  0.02   NaN   NaN  NaN  NaN  0.00   NaN  
3  ...    NaN  0.02  0.39  0.28   NaN   NaN  NaN  NaN  0.00   NaN  
4  ...    NaN  0.04  0.09  0.02   NaN   NaN  NaN  NaN  0.00   NaN  
5  ...    NaN  0.01  0.58  0.10   NaN   NaN  NaN  NaN  0.00   NaN  
6  ...    NaN  0.05  0.08  0.06   NaN   NaN  NaN  NaN  0.00   NaN  
7  ...    NaN  0.01  0.33  0.00   NaN   NaN  NaN  NaN  0.00   NaN  
8  ...    NaN  0.04  0.17  0.04   NaN   NaN  NaN  NaN  0.00   NaN  
9  ...    NaN  0.00  0.47  0.11   NaN   NaN  NaN  NaN  0.00   NaN  

[10 rows x 128 columns]

## EDA

In [34]:
print(crime.shape)
print(crime.dtypes)
print(crime.mean())
print()
print(crime.std())

crime_community=pd.DataFrame(crime.groupby(3).size())
# Since there are few communicy names that are duplicated we can drop this categorical variable.  


(1994, 128)
127    float64
0        int64
1      float64
2      float64
3       object
4        int64
5      float64
6      float64
7      float64
8      float64
9      float64
10     float64
11     float64
12     float64
13     float64
14     float64
15     float64
16     float64
17     float64
18     float64
19     float64
20     float64
21     float64
22     float64
23     float64
24     float64
25     float64
26     float64
27     float64
28     float64
        ...   
97     float64
98     float64
99     float64
100    float64
101    float64
102    float64
103    float64
104    float64
105    float64
106    float64
107    float64
108    float64
109    float64
110    float64
111    float64
112    float64
113    float64
114    float64
115    float64
116    float64
117    float64
118    float64
119    float64
120    float64
121    float64
122    float64
123    float64
124    float64
125    float64
126    float64
Length: 128, dtype: object
127        0.237979
0         28.683551
1     

In [30]:
## Removing categorical community name.  
crime2=crime.drop(columns=[3])
crime2.head(10)
crime2.shape

(1994, 127)

In [45]:
# Data is full of missing values in columns: 1, 2, 101:117, 121:124 ,126
# Because the majority of data is incomplete from these columns it is not worth implementing a mean/median.  
## Adding this in would likely just make the data less accurate overall.  
crime3=crime2.drop(columns=[1,2,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,121,122,123,124,126])
crime3.shape

(1994, 103)

In [64]:
#Removing last NA in dataset
crime4 = crime3.dropna()
crime4.shape

## Test/Train 70/30

In [66]:
features_train, features_test, target_train, target_test = train_test_split(
    crime4.iloc[:,1:].values, crime4.iloc[:,0].values, test_size=0.30, random_state=0)

In [90]:
print(features_test.shape)
print(features_train.shape)
print(target_test.shape)
print(target_train.shape)

(598, 102)
(1395, 102)
(598,)
(1395,)


## Ridge

In [70]:
from sklearn.linear_model import Ridge
ridge_reg = Ridge(alpha=0, normalize=True)
ridge_reg.fit(features_train, target_train)
target_predicted_ridge = ridge_reg.predict(features_test)


## Print Coefficients/Metrics

In [81]:
print(ridge_reg.coef_)
mse = np.mean((target_predicted_ridge-target_test)**2)
print('mse')
print(mse)
## Explained Variance Score
from sklearn.metrics import explained_variance_score
print('explained variance score')
print(explained_variance_score(target_test, target_predicted_ridge))  

from sklearn.metrics import mean_absolute_error
print('mean absolute error')
print(mean_absolute_error(target_test, target_predicted_ridge))


[-0.00120101 -0.00122226 -0.14885853 -0.0183277   0.19325857 -0.0122058
 -0.00662831  0.06644441  0.16996948 -0.37812796 -0.15824475 -0.11775627
  0.05824389  0.04317562 -0.41819672 -0.16482448  0.05426248 -0.15287171
  0.20750527 -0.0374141  -0.10568712  0.30453586  0.05998656 -0.22447831
 -0.00211392 -0.0347317   0.00318169  0.0474694   0.05016545 -0.0074569
 -0.16654445 -0.09420076  0.03954075  0.07970864  0.03302132  0.211794
 -0.05756007 -0.04571954  0.04489486  0.09794207  0.52540464  0.23553379
  0.08531321 -0.57377329 -0.07866041 -0.05631369 -0.28709949 -0.02070768
  0.03893017  0.03614124 -0.16786376 -0.10603024  0.14611987 -0.0913311
  0.07531689 -0.07225578  0.01584703  0.01138706 -0.0785271   0.02838397
 -0.08226601  0.20081308 -0.02352697 -0.15978351 -0.28409524  0.12764193
  0.7903841  -0.27025235 -0.29437171 -0.51621912  0.19547768  0.04164927
  0.01628799  0.19518775 -0.04248437  0.29686165  0.06560693 -0.05855956
 -0.01018869  0.00816236 -0.00812833 -0.22996698  0.1940

## LASSO Regression

In [85]:
from sklearn.linear_model import Lasso
lassoreg = Lasso(alpha=0.1)
lassoreg.fit(features_train, target_train)
target_predicted_lasso = lassoreg.predict(features_test)


In [87]:
print(lassoreg.coef_)
mse = np.mean((target_predicted_lasso-target_test)**2)
print('mse')
print(mse)
## Explained Variance Score
from sklearn.metrics import explained_variance_score
print('explained variance score')
print(explained_variance_score(target_test, target_predicted_lasso))  

from sklearn.metrics import mean_absolute_error
print('mean absolute error')
print(mean_absolute_error(target_test, target_predicted_lasso))

[-0.00272227 -0.          0.         -0.          0.         -0.
 -0.          0.          0.          0.          0.          0.
  0.          0.         -0.         -0.         -0.         -0.
  0.          0.         -0.         -0.         -0.         -0.
 -0.         -0.         -0.         -0.         -0.          0.
  0.          0.          0.         -0.          0.         -0.
 -0.         -0.          0.         -0.          0.          0.
  0.          0.          0.         -0.         -0.         -0.
 -0.         -0.         -0.          0.          0.          0.
  0.          0.          0.          0.          0.          0.
  0.          0.         -0.          0.          0.          0.
 -0.         -0.          0.         -0.          0.          0.
 -0.          0.         -0.         -0.          0.          0.
 -0.          0.          0.         -0.         -0.         -0.
 -0.         -0.         -0.         -0.          0.         -0.
  0.          0.         

## Elastic Net Regression

In [95]:
from sklearn.linear_model import ElasticNet
en = ElasticNet(alpha=0.1)
en.fit(features_train, target_train)
print(en)
target_predicted_en = en.predict(features_test)


ElasticNet(alpha=0.1, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=None, selection='cyclic', tol=0.0001, warm_start=False)


In [96]:
print(en.coef_)
mse = np.mean((target_predicted_en-target_test)**2)
print('mse')
print(mse)
## Explained Variance Score
from sklearn.metrics import explained_variance_score
print('explained variance score')
print(explained_variance_score(target_test, target_predicted_en))  

from sklearn.metrics import mean_absolute_error
print('mean absolute error')
print(mean_absolute_error(target_test, target_predicted_en))

[-0.00290628 -0.          0.         -0.          0.         -0.
 -0.          0.          0.          0.          0.          0.
  0.          0.         -0.         -0.         -0.         -0.
  0.          0.         -0.         -0.         -0.         -0.
 -0.         -0.         -0.         -0.         -0.          0.
  0.          0.          0.         -0.          0.         -0.
 -0.         -0.          0.         -0.          0.          0.
  0.          0.          0.         -0.         -0.         -0.
 -0.         -0.         -0.          0.          0.          0.
  0.          0.          0.          0.          0.          0.
  0.          0.         -0.          0.          0.          0.
 -0.         -0.          0.         -0.          0.          0.
 -0.          0.         -0.         -0.          0.          0.
 -0.          0.          0.         -0.         -0.         -0.
 -0.         -0.         -0.         -0.          0.         -0.
  0.          0.         

## Ordinary Least Squares

In [113]:
from sklearn import linear_model
reg = linear_model.LinearRegression()
ols = reg.fit (features_train, target_train)
target_predicted_reg = reg.predict(features_test)

In [114]:
print(reg.coef_)
mse = np.mean((target_predicted_reg-target_test)**2)
print('mse')
print(mse)
## Explained Variance Score
from sklearn.metrics import explained_variance_score
print('explained variance score')
print(explained_variance_score(target_test, target_predicted_reg))  

from sklearn.metrics import mean_absolute_error
print('mean absolute error')
print(mean_absolute_error(target_test, target_predicted_reg))

[-0.00120101 -0.00122226 -0.14885853 -0.0183277   0.19325857 -0.0122058
 -0.00662831  0.06644441  0.16996948 -0.37812796 -0.15824475 -0.11775627
  0.05824389  0.04317562 -0.41819672 -0.16482448  0.05426248 -0.15287171
  0.20750527 -0.0374141  -0.10568712  0.30453586  0.05998656 -0.22447831
 -0.00211392 -0.0347317   0.00318169  0.0474694   0.05016545 -0.0074569
 -0.16654445 -0.09420076  0.03954075  0.07970864  0.03302132  0.211794
 -0.05756007 -0.04571954  0.04489486  0.09794207  0.52540464  0.23553379
  0.08531321 -0.57377329 -0.07866041 -0.05631369 -0.28709949 -0.02070768
  0.03893017  0.03614124 -0.16786376 -0.10603024  0.14611987 -0.0913311
  0.07531689 -0.07225578  0.01584703  0.01138706 -0.0785271   0.02838397
 -0.08226601  0.20081308 -0.02352697 -0.15978351 -0.28409524  0.12764193
  0.7903841  -0.27025235 -0.29437171 -0.51621912  0.19547768  0.04164927
  0.01628799  0.19518775 -0.04248437  0.29686165  0.06560693 -0.05855956
 -0.01018869  0.00816236 -0.00812833 -0.22996698  0.1940

## Part B - Grid Search

In [101]:
# use a full grid over several parameters and cross validate 5 times
from sklearn.model_selection import GridSearchCV
#param_grid = {"alpha": [.01,.1, .5, 1, 2]}
#param_grid={"alpha": [1,10,1]} this does a range 1 through 10 changes by a factor of 1. 
param_grid={"alpha": [.01,1,.05]} #this does a range 1 through 1 changes by a factor of .05

# run grid search
grid_search = GridSearchCV(ridge_reg, param_grid=param_grid,n_jobs=-1,cv=5)
grid_search.fit(features_train, target_train)
print("Grid Scores", grid_search.cv_results_)
print("Best", grid_search.best_params_)    

Grid Scores {'mean_fit_time': array([0.01211562, 0.0050005 , 0.00435753]), 'std_fit_time': array([0.0032142 , 0.00076641, 0.00087521]), 'mean_score_time': array([0.00108976, 0.00057673, 0.00048251]), 'std_score_time': array([4.17865420e-04, 1.04912302e-04, 5.62977706e-05]), 'param_alpha': masked_array(data=[0.01, 1, 0.05],
             mask=[False, False, False],
       fill_value='?',
            dtype=object), 'params': [{'alpha': 0.01}, {'alpha': 1}, {'alpha': 0.05}], 'split0_test_score': array([0.65801934, 0.63362382, 0.65457334]), 'split1_test_score': array([0.57585792, 0.57941544, 0.5828878 ]), 'split2_test_score': array([0.71404975, 0.68572192, 0.71377524]), 'split3_test_score': array([0.63750359, 0.64130553, 0.65111674]), 'split4_test_score': array([0.65858524, 0.64310757, 0.66006382]), 'mean_test_score': array([0.64880317, 0.63663486, 0.65248339]), 'std_test_score': array([0.04446669, 0.03393046, 0.04162676]), 'rank_test_score': array([2, 3, 1], dtype=int32), 'split0_train_sco

In [103]:
# use a full grid over several parameters and cross validate 5 times
from sklearn.model_selection import GridSearchCV
#param_grid = {"alpha": [.01,.1, .5, 1, 2]}
#param_grid={"alpha": [1,10,1]} this does a range 1 through 10 changes by a factor of 1. 
param_grid={"alpha": [.01,1,.05]} #this does a range 1 through 1 changes by a factor of .05

# run grid search
grid_search = GridSearchCV(lassoreg, param_grid=param_grid,n_jobs=-1,cv=5)
grid_search.fit(features_train, target_train)
print("Grid Scores", grid_search.cv_results_)
print("Best", grid_search.best_params_)   

Grid Scores {'mean_fit_time': array([0.01457019, 0.00629325, 0.0038475 ]), 'std_fit_time': array([0.00120691, 0.00162928, 0.00089793]), 'mean_score_time': array([0.0024497 , 0.00065751, 0.00054522]), 'std_score_time': array([1.15877276e-03, 9.55242298e-05, 5.47730870e-05]), 'param_alpha': masked_array(data=[0.01, 1, 0.05],
             mask=[False, False, False],
       fill_value='?',
            dtype=object), 'params': [{'alpha': 0.01}, {'alpha': 1}, {'alpha': 0.05}], 'split0_test_score': array([ 0.55927174, -0.00083497,  0.06484973]), 'split1_test_score': array([ 0.53330187, -0.00120038,  0.05673995]), 'split2_test_score': array([ 0.57733858, -0.00415995,  0.03653142]), 'split3_test_score': array([ 0.57782429, -0.00122184,  0.02708167]), 'split4_test_score': array([ 0.53485568, -0.00096124,  0.04832709]), 'mean_test_score': array([ 0.55651843, -0.00167568,  0.04670597]), 'std_test_score': array([0.01951042, 0.00125065, 0.01357306]), 'rank_test_score': array([1, 3, 2], dtype=int32),

In [105]:
# use a full grid over several parameters and cross validate 5 times
from sklearn.model_selection import GridSearchCV
#param_grid = {"alpha": [.01,.1, .5, 1, 2]}
#param_grid={"alpha": [1,10,1]} this does a range 1 through 10 changes by a factor of 1. 
param_grid={"alpha": [.01,1,.05]} #this does a range 1 through 1 changes by a factor of .05

# run grid search
grid_search = GridSearchCV(en, param_grid=param_grid,n_jobs=-1,cv=5)
grid_search.fit(features_train, target_train)
print("Grid Scores", grid_search.cv_results_)
print("Best", grid_search.best_params_)   

Grid Scores {'mean_fit_time': array([0.01249485, 0.00475941, 0.00395207]), 'std_fit_time': array([0.00203116, 0.00060102, 0.00073038]), 'mean_score_time': array([0.00132585, 0.00058446, 0.00053363]), 'std_score_time': array([4.30618356e-04, 4.87419015e-05, 4.46008203e-05]), 'param_alpha': masked_array(data=[0.01, 1, 0.05],
             mask=[False, False, False],
       fill_value='?',
            dtype=object), 'params': [{'alpha': 0.01}, {'alpha': 1}, {'alpha': 0.05}], 'split0_test_score': array([0.60546248, 0.03509889, 0.30791149]), 'split1_test_score': array([0.57880124, 0.03397061, 0.30903576]), 'split2_test_score': array([0.64460764, 0.02233251, 0.26252923]), 'split3_test_score': array([0.63137662, 0.02573938, 0.29676883]), 'split4_test_score': array([0.58819165, 0.02990463, 0.28556954]), 'mean_test_score': array([0.60968793, 0.0294092 , 0.29236297]), 'std_test_score': array([0.02499487, 0.00483677, 0.01717762]), 'rank_test_score': array([1, 3, 2], dtype=int32), 'split0_train_sco

## Re-Run Lasso and EN with optimized Alpha

In [106]:
from sklearn.linear_model import Lasso
lassoreg = Lasso(alpha=0.01)
lassoreg.fit(features_train, target_train)
target_predicted_lasso = lassoreg.predict(features_test)


In [107]:
print(lassoreg.coef_)
mse = np.mean((target_predicted_lasso-target_test)**2)
print('mse')
print(mse)
## Explained Variance Score
from sklearn.metrics import explained_variance_score
print('explained variance score')
print(explained_variance_score(target_test, target_predicted_lasso))  

from sklearn.metrics import mean_absolute_error
print('mean absolute error')
print(mean_absolute_error(target_test, target_predicted_lasso))

[-0.0020639  -0.00054276  0.         -0.          0.         -0.09798879
  0.          0.         -0.         -0.         -0.          0.
  0.          0.         -0.         -0.         -0.         -0.
  0.          0.         -0.         -0.         -0.         -0.
 -0.         -0.         -0.         -0.         -0.          0.
  0.          0.          0.         -0.          0.         -0.
 -0.         -0.          0.         -0.          0.          0.
  0.          0.          0.         -0.         -0.21921124 -0.
 -0.         -0.         -0.          0.          0.27544601  0.
  0.          0.          0.          0.          0.          0.
  0.          0.         -0.          0.          0.          0.
 -0.         -0.          0.         -0.          0.          0.
 -0.          0.         -0.         -0.          0.          0.
 -0.          0.          0.         -0.         -0.         -0.
 -0.         -0.         -0.         -0.          0.          0.
 -0.          0. 

In [108]:
from sklearn.linear_model import ElasticNet
en = ElasticNet(alpha=0.01)
en.fit(features_train, target_train)
print(en)
target_predicted_en = en.predict(features_test)


ElasticNet(alpha=0.01, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=None, selection='cyclic', tol=0.0001, warm_start=False)


In [109]:
print(en.coef_)
mse = np.mean((target_predicted_en-target_test)**2)
print('mse')
print(mse)
## Explained Variance Score
from sklearn.metrics import explained_variance_score
print('explained variance score')
print(explained_variance_score(target_test, target_predicted_en))  

from sklearn.metrics import mean_absolute_error
print('mean absolute error')
print(mean_absolute_error(target_test, target_predicted_en))

[-0.00198478 -0.00112982  0.         -0.          0.0397667  -0.1294153
  0.          0.         -0.         -0.         -0.          0.
  0.          0.02342869 -0.         -0.         -0.         -0.
  0.          0.         -0.         -0.         -0.         -0.
 -0.         -0.         -0.          0.         -0.          0.
  0.          0.          0.         -0.          0.         -0.
 -0.         -0.          0.         -0.          0.03966757  0.
  0.          0.01667577  0.         -0.05770051 -0.14274881 -0.05038667
 -0.         -0.         -0.          0.          0.18720762  0.
  0.          0.          0.          0.          0.          0.
  0.          0.         -0.          0.          0.          0.
 -0.         -0.          0.         -0.          0.01463211  0.
 -0.          0.02443093 -0.00152117 -0.          0.05499071 -0.
 -0.          0.01425173  0.         -0.         -0.         -0.
 -0.         -0.         -0.         -0.          0.          0.
 -0.      

## Results LASSO/EN
Both Lasso and Elastic Net had almost identical results throughout.  I believe this is because EN is a function of Ridge and Lasso, and Lasso must dominate that function.  Although the difference between Alpha = 0.1 and (optimized thorugh grid search) Alpha = 0.01 was significant bring the MSE down from 0.05 to 0.02, the MAE from 0.18 to 0.11 and the Explained Variance Score from only 4% to about 56%.  A considerable difference, clearly showing the value of grid search.  

## Part C Cross Validation OLS
Cross Validation suggests the OLS model is not close to optimized and that the variance is not captured well in the model training.  

## Results OLS
Since Alpha = 0 for Ridge Regression is equivalent to OLS there is no difference between the initial Ridge model and this OLS model.  The Cross Validation suggests the model is not stable enough across the dataset to be production ready.  

In [115]:
#verify OLS with Cross Validation
scores = cross_val_score(ols, features_train, target_train, cv=10)
print("Cross Validation Score for each K",scores)
scores.mean()   

Cross Validation Score for each K [0.652178   0.65726692 0.69227167 0.44356508 0.63588469 0.77090299
 0.60950888 0.62971867 0.61099369 0.70503524]


0.6407325834942149

## Cross Validation RIDGE

In [116]:
#verify Ridge with Cross Validation
scores = cross_val_score(ridge_reg, features_train, target_train, cv=10)
print("Cross Validation Score for each K",scores)
scores.mean()   

Cross Validation Score for each K [0.652178   0.65726692 0.69227167 0.44356508 0.63588469 0.77090299
 0.60950888 0.62971867 0.61099369 0.70503524]


0.6407325834942128

## Optimized RIDGE

In [119]:
from sklearn.linear_model import Ridge
ridge_reg = Ridge(alpha=0.05, normalize=True)
ridge_reg.fit(features_train, target_train)
target_predicted_ridge = ridge_reg.predict(features_test)


In [120]:
print(ridge_reg.coef_)
mse = np.mean((target_predicted_ridge-target_test)**2)
print('mse')
print(mse)
## Explained Variance Score
from sklearn.metrics import explained_variance_score
print('explained variance score')
print(explained_variance_score(target_test, target_predicted_ridge))  

from sklearn.metrics import mean_absolute_error
print('mean absolute error')
print(mean_absolute_error(target_test, target_predicted_ridge))


[-0.001267   -0.00132035 -0.04933722  0.04764059  0.13742705 -0.07322984
 -0.02210096  0.02328524  0.03413549 -0.18453991 -0.04156441  0.00664928
 -0.03244781  0.04277883 -0.00097864 -0.06693557  0.03116259 -0.10440764
  0.050757   -0.01315657 -0.0941628   0.00469434 -0.02005108 -0.04741583
 -0.00636212 -0.030836    0.00949226  0.04554592  0.04089317 -0.0133442
 -0.06911149 -0.04691693  0.01034636  0.01582683 -0.01404344  0.06900133
 -0.0308182  -0.03545605  0.01908679  0.02264101  0.12454445  0.08719826
 -0.07039577 -0.00163717  0.03668005 -0.07924686 -0.11091736 -0.05591899
  0.00208114 -0.00384911 -0.09724429 -0.03840525  0.15136967 -0.09114554
  0.04256771 -0.03627243 -0.01544407  0.01246744 -0.00331183  0.00757354
  0.02100546  0.04550414  0.01730584 -0.04390395 -0.04102409 -0.02146219
  0.10045115 -0.07743653 -0.01923619 -0.08245864  0.08940299  0.02483394
  0.00040034  0.13731275 -0.0618891  -0.02681386  0.06939625 -0.03570318
  0.00032867  0.01460538 -0.00800227 -0.03378898 -0.

In [121]:
#verify Ridge with Cross Validation
scores = cross_val_score(ridge_reg, features_train, target_train, cv=10)
print("Cross Validation Score for each K",scores)
scores.mean()   

Cross Validation Score for each K [0.65208075 0.65637244 0.69501219 0.4635235  0.64090128 0.77307429
 0.66490639 0.64537331 0.61338665 0.7106114 ]


0.6515242210817016

## Results/Conclusions RIDGE (Optimized Model)
Initial Results:
Ridge Alpha = 0 
- MSE - .019
- EVS - .6628
- MAE - .096
Grid Search Results 
- Alpha = 0.05
Ridge Alpha = 0.05
- MSE - .0189
- EVS - .668
- MAE - .094

Prior to starting I did not realize that because Alpha=0 for my initial Ridge Regression there would be no difference to OLS.  Of course, after doing both models I realized this and after running a grid search through .1-2 I ran through .01-.05 and found the optimized penalty to be alpha = 0.05.  This is the penalty multiplier that reduces bias and variance the most build the most accurate and stable model.  However during CV I found there were considerable outliers suggesting the full variance of the dataset is not captured during training.  

From Alpha = 0 to Alpha = 0.05 the MSE, EVS, and MAE improved very minimally, suggesting that the penalty method may not be necessary in this solution.  The MAE evaluates the absolute difference in error of the means.  It is easily interpretable.  The explained variance score states how well the variance describes the prediction results.  In this case 66% is not great, but not terrible.  This is relatable to the Cross Validation Score.  I'd expect higher CV score variation if the EVS gets lower.  
